subprojects/file-collections/src/main/java/org/gradle/api/internal/file/CompositeFileTree.java

```diff
@@ -50,20 +55,20 @@ protected List<? extends FileTreeInternal> getSourceCollections() {
 
     @Override
     public FileTree plus(FileTree fileTree) {
-        return new UnionFileTree(this, Cast.cast(FileTreeInternal.class, fileTree));
+        return new UnionFileTree(taskDependencyFactory, this, Cast.cast(FileTreeInternal.class, fileTree));
     }
 
     @Override
     public FileTree matching(final Closure filterConfigClosure) {
-        return new FilteredFileTree(this, patternSetFactory, () -> {
+        return new FilteredFileTree(this, taskDependencyFactory, patternSetFactory, () -> {
```

Comment:
:art: Remove duplication then inject the `taskDependencyFactory` in a single place

```java
    @Override
    public FileTree matching(final Closure filterConfigClosure) {
        return newFilteredFileTree(() -> {
           ... 
        });
    }

    @Override
    public FileTree matching(final Action<? super PatternFilterable> filterConfigAction) {
        return newFilteredFileTree(() -> {
           ... 
        });
    }

    @Override
    public FileTreeInternal matching(final PatternFilterable patterns) {
        return newFilteredFileTree(() -> {
           ... 
        });
    }

    private FilteredFileTree newFilteredFileTree(Supplier<PatternSet> patternSetSupplier) {
        return new FilteredFileTree(this, taskDependencyFactory, patternSetFactory, patternSetSupplier);
    }
``

Reply:
✅ in https://github.com/gradle/gradle/pull/22433/commits/5d25045538a90f672caf3c7f9e5666e6964a6ebd